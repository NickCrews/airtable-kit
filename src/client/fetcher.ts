import console from "node:console";
import process from "node:process";
export type Fetcher = {
    fetch: (args: {
        path: string;
        method?: "GET" | "POST" | "PATCH" | "PUT" | "DELETE";
        data?: unknown;
    }) => Promise<unknown>;
};

export type FetcherArgs = {
    apiKey?: string;
    baseUrl?: string;
};

export type IntoFetcher = Fetcher | string | FetcherArgs | undefined;

export function createRealFetcher(args?: IntoFetcher): Fetcher {
    if (typeof args === "string") {
        return createRealFetcher({ apiKey: args });
    }
    if (args && "fetch" in args) {
        return args;
    }
    const baseUrl = args?.baseUrl ?? "https://api.airtable.com/v0";
    let apiKey = args?.apiKey;
    if (!apiKey) {
        apiKey = process.env.AIRTABLE_API_KEY;
    }
    return {
        fetch: async ({ path, method = "GET", data }) => {
            path = path.startsWith("/") ? path.slice(1) : path;
            const response = await fetch(`${baseUrl}/${path}`, {
                method,
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    "Content-Type": "application/json",
                },
                body: JSON.stringify(data),
            });
            return response.json();
        },
    };
}

export function createMockFetcher(): Fetcher & {
    setReturnValue: (value: unknown) => void;
    getCallHistory: () => unknown[];
    reset: () => void;
} {
    let returnValue: unknown = undefined;
    let callHistory: unknown[] = [];
    function mockFetch(
        args: { path: string; method?: string; data?: unknown },
    ): Promise<unknown> {
        console.log(
            "Mock fetch called with args:",
            args,
        );
        callHistory.push(args);
        return Promise.resolve(returnValue);
    }
    return {
        fetch: mockFetch,
        setReturnValue: (value: unknown) => {
            returnValue = value;
        },
        getCallHistory: () => callHistory,
        reset: () => {
            callHistory = [];
        },
    };
}
