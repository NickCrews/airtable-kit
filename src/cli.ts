#!/usr/bin/env node
/**
 * CLI for airtable-kit
 */

import { parseArgs } from 'node:util';
import { BaseId, fetchBaseSchema } from "./index.ts";
import { generateCode } from "./codegen/index.ts";
import { promises as fs } from "node:fs";
import * as path from "node:path";
import { IntoFetcher } from './client/fetcher.ts';

import * as packageJson from '../package.json';

async function doCodegen({
  baseId,
  fetcher,
  outPath,
  baseName,
}: {
  baseId: BaseId;
  outPath: string;
  baseName?: string;
  fetcher: IntoFetcher;
}, console: Consolish = globalThis.console): Promise<void> {
  console.log("üîç Fetching schema from Airtable...");
  const resolvedBaseName = getBaseName(outPath, baseName);
  const baseSchema = await fetchBaseSchema({ baseId, fetcher, baseName: resolvedBaseName });
  const firstTable = baseSchema.tables[0];

  console.log(`‚úì Found base with ${baseSchema.tables.length} tables`);
  console.log(`  Using Base Name: ${baseSchema.name}`);
  console.log(`  Tables:\n${baseSchema.tables.map((t) => t.name).join("\n    - ")}`);

  const filetype = outPath.endsWith(".ts") ? "ts" : "js";
  const code = generateCode(baseSchema, { filetype });

  const outputDir = path.dirname(outPath);
  await fs.mkdir(outputDir, { recursive: true });
  await fs.writeFile(outPath, code, "utf-8");

  console.log(`\n‚úÖ Generated schema at ${outPath}`);
  console.log("\nüìö Example usage:");
  const importPath = outPath.startsWith("/") ? outPath : outPath.startsWith(".") ? outPath : `./${outPath}`;
  console.log(
    `import myBaseSchema from '${importPath}';
import { baseClient } from 'airtable-kit/client';

const client = baseClient({
  baseSchema: myBaseSchema,
  fetcher: 'YOUR_API_KEY',
});
const records = await client.tables.${firstTable.name}.list()
console.log(records);`
  );
}

function getHelp() {
  return `Airtable Kit CLI

Usage:
  npx airtable-kit codegen --base-id <BASE_ID> --api-key <API_KEY> --output <OUTPUT_PATH> [--base-name <BASE_NAME>]
  npx airtable-kit --help
  npx airtable-kit --version

Commands:
  codegen          Generate TypeScript or JavaScript schema file for an Airtable base.

Options:
  --base-id       The ID of the Airtable base to generate the schema from. (required)
  --api-key       Your Airtable API key or a custom fetcher to use for fetching the schema. (required)
  --output        The output path for the generated schema file (e.g., schemas/myBase.ts). (required)
                  Can be a .ts or .js file, and the output format will be inferred accordingly.
  --base-name     Optional custom name for the base in the generated schema.
                  If not provided, will be inferred from the output filename.
  
  --help          Show this help message.
  --version       Show the version number.
`.trim();
}

type Consolish = {
  log: (...args: unknown[]) => void;
  error: (...args: unknown[]) => void;
};

function getBaseName(outPath: string, baseName?: string): string {
  if (baseName) {
    return baseName;
  }
  const filename = path.basename(outPath);
  return filename.split(".")[0];
}

export async function cli(args: string[], fetcher?: IntoFetcher, console: Consolish = globalThis.console): Promise<void> {
  const options = {
    version: { type: 'boolean' },
    help: { type: 'boolean' },
    "base-id": { type: 'string' },
    "api-key": { type: 'string' },
    "base-name": { type: 'string' },
    output: { type: 'string' },
  } as const;
  const { values, positionals } = parseArgs({ args, options, allowPositionals: true });

  if (values.help || positionals[0] === 'help') {
    console.log(getHelp());
    return;
  }
  if (values.version || positionals[0] === 'version') {
    console.log(packageJson.version);
    return;
  }
  if (positionals[0] === 'codegen') {
    const missings = [];
    if (!values["base-id"]) {
      missings.push('--base-id');
    }
    if (!values["api-key"]) {
      missings.push('--api-key');
    }
    if (!values.output) {
      missings.push('--output');
    }
    if (missings.length > 0) {
      const err = `Missing required options: ${missings.join(', ')}`;
      console.error(`‚ùå ${err}`);
      console.log(getHelp());
      throw new Error(err);
    }
    await doCodegen({
      baseId: values["base-id"] as BaseId,
      fetcher: fetcher ?? values["api-key"] as string,
      outPath: values.output as string,
      baseName: values["base-name"],
    }, console);
    return;
  }
  console.log(getHelp());
  throw new Error(`Unknown command`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  cli(process.argv.slice(2)).catch((error) => {
    console.error("‚ùå Error:", (error as Error).message);
    process.exit(1);
  });
}
