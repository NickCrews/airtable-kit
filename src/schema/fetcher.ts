import { type TableSchema } from "./tables.js";
import { type BaseSchema } from "./bases.js";
import { type IntoFetcher, makeFetcher } from "../client/fetcher.js";

export async function fetchBaseSchema(
  baseId: string,
  fetcher: IntoFetcher,
): Promise<Omit<BaseSchema, "name">> {
  const path = `/meta/bases/${baseId}/tables`;
  const realFetcher = makeFetcher(fetcher);
  const response = await realFetcher.fetch({ path });

  return {
    id: baseId as `app${string}`,
    tables: (response as { tables: unknown[] }).tables as TableSchema[],
  };
}
