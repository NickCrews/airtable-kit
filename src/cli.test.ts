/**
 * Tests for CLI commands
 */

import fs from 'node:fs';
import { describe, it, expect, beforeEach, vi } from 'vitest';
import tasksSchema from './tests/taskBase.ts';
import { type Fetcher } from './client/fetcher.ts';
import { cli } from './cli.ts';
import { makeInTmpDir } from './tests/inTmpDir.ts';

describe('CLI', () => {
  const mockFetch = vi.fn(async (args: { path: string }) => {
    if (args.path === '/meta/bases') {
      return {
        bases: [
          { id: 'appOne', name: 'Task Base', permissionLevel: 'edit' },
          { id: 'appTwo', name: 'Other Base', permissionLevel: 'edit' },
        ],
      };
    }
    if (args.path === '/meta/bases/appOne/tables' || args.path === '/meta/bases/appTwo/tables') {
      return { tables: (tasksSchema as any).tables };
    }
    if (args.path.startsWith('/meta/bases/')) {
      return { tables: (tasksSchema as any).tables };
    }
    return tasksSchema as any;
  });
  const mockConsole = {
    log: vi.fn(),
    error: vi.fn(),
  }
  const mockFetcher = { fetch: mockFetch } as Fetcher;

  const mockedCli = (args: string[]) => cli(args, mockFetcher, mockConsole);

  beforeEach(() => {
    vi.clearAllMocks();
  });

  const inTmpDir = makeInTmpDir();

  describe('codegen command', () => {
    it('should create output directory', async (ctx) => {
      const tmpDir = inTmpDir(ctx);
      await mockedCli(["codegen", "base", "appTestBase", "--api-key", "patTest123", "--outfile", "schemas/frEvents.ts"]);
      expect(mockFetch).toHaveBeenCalledWith({ path: '/meta/bases/appTestBase/tables' });
      expect(fs.readFileSync(`${tmpDir}/schemas/frEvents.ts`, 'utf-8')).toMatchSnapshot();
    });

    it('should generate  js', async (ctx) => {
      const tmpDir = inTmpDir(ctx);
      await mockedCli(["codegen", "base", "appTestBase", "--api-key", "patTest123", "--outfile", "schemas/frEvents.js"]);
      expect(mockFetch).toHaveBeenCalledWith({ path: '/meta/bases/appTestBase/tables' });
      expect(fs.readFileSync(`${tmpDir}/schemas/frEvents.js`, 'utf-8')).toMatchSnapshot();
    });

    it('should use custom --base-name', async (ctx) => {
      const tmpDir = inTmpDir(ctx);
      await mockedCli(["codegen", "base", "appTestBase", "--api-key", "patTest123", "--outfile", "schemas/frEvents.js", "--base-name", "My Custom Base"]);
      expect(mockFetch).toHaveBeenCalledWith({ path: '/meta/bases/appTestBase/tables' });
      expect(fs.readFileSync(`${tmpDir}/schemas/frEvents.js`, 'utf-8')).toMatchSnapshot();
    });

    it('should default outfile to ./<BASE_NAME>-schema.ts when omitted', async (ctx) => {
      const tmpDir = inTmpDir(ctx);
      await mockedCli(["codegen", "base", "appTestBase", "--api-key", "patTest123"]);
      expect(mockFetch).toHaveBeenCalledWith({ path: '/meta/bases/appTestBase/tables' });
      expect(fs.readFileSync(`${tmpDir}/appTestBase-schema.ts`, 'utf-8')).toMatchSnapshot();
    });

    it('should generate all base schemas into outdir', async (ctx) => {
      const tmpDir = inTmpDir(ctx);
      await mockedCli(["codegen", "all", "--api-key", "patTest123", "--outdir", "schemas", "--format", "ts"]);
      expect(fs.readFileSync(`${tmpDir}/schemas/taskBase-schema.ts`, 'utf-8')).toMatchSnapshot();
      expect(fs.existsSync(`${tmpDir}/schemas/otherBase-schema.ts`)).toBe(true);
    });
  });

  describe('version and help', () => {
    it('should display help text with "--help"', async () => {
      await mockedCli(["--help"]);
      expect(mockConsole.log).toHaveBeenCalledWith(expect.stringContaining('Airtable Kit CLI'));
      expect(mockFetch).not.toHaveBeenCalled();
    });

    it('should display help text with "help"', async () => {
      await mockedCli(["help"]);
      expect(mockConsole.log).toHaveBeenCalledWith(expect.stringContaining('Airtable Kit CLI'));
      expect(mockFetch).not.toHaveBeenCalled();
    });

    it('should display version with "--version"', async () => {
      await mockedCli(["--version"]);
      expect(mockConsole.log).toHaveBeenCalledWith(expect.stringMatching(/^\d+\.\d+\.\d+$/));
      expect(mockFetch).not.toHaveBeenCalled();
    });
    it('should display version with "version"', async () => {
      await mockedCli(["version"]);
      expect(mockConsole.log).toHaveBeenCalledWith(expect.stringMatching(/^\d+\.\d+\.\d+$/));
      expect(mockFetch).not.toHaveBeenCalled();
    });
  });

  describe('CLI error handling', () => {
    it('should require base-id for codegen base command', async () => {
      await expect(
        async () => await mockedCli(["codegen", "base", "--api-key", "patTest123"])
      ).rejects.toThrow();
    });

    it('should require api-key for codegen base command', async () => {
      await expect(
        async () => await mockedCli(["codegen", "base", "appTestBase"])
      ).rejects.toThrow();
    });
  });
});
