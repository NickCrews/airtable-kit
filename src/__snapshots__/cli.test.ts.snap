// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CLI > codegen command > should create output directory 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTestBase",
  "name": "appTestBase",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "Name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "Status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "Priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "Due Date",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "Completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "Notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "Tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "Attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "Created At",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "Updated At",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "Assigned To",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "Email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "Full Name",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "Phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
} as const;
"
`;

exports[`CLI > codegen command > should generate  js 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTestBase",
  "name": "appTestBase",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "Name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "Status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "Priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "Due Date",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "Completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "Notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "Tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "Attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "Created At",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "Updated At",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "Assigned To",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "Email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "Full Name",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "Phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
};
"
`;

exports[`CLI > codegen command > should use custom --base-name 1`] = `
"/**
 * Auto-generated from Airtable schema
 * Do not edit manually
 */

export default {
  "id": "appTestBase",
  "name": "My Custom Base",
  "tables": [
    {
      "id": "tblTasks",
      "name": "tasks",
      "fields": [
        {
          "id": "fldName",
          "name": "Name",
          "type": "singleLineText",
          "description": "The name of the task"
        },
        {
          "id": "fldStatus",
          "name": "Status",
          "type": "singleSelect",
          "description": "The current status of the task",
          "options": {
            "choices": [
              {
                "id": "selTodo",
                "name": "Todo",
                "color": "blueLight2"
              },
              {
                "id": "selInProgress",
                "name": "In Progress",
                "color": "yellowLight2"
              },
              {
                "id": "selDone",
                "name": "Done",
                "color": "greenLight2"
              }
            ]
          }
        },
        {
          "id": "fldPriority",
          "name": "Priority",
          "type": "number",
          "description": "The priority level of the task. Lower numbers indicate higher priority.",
          "options": {
            "precision": 0
          }
        },
        {
          "id": "fldDueDate",
          "name": "Due Date",
          "type": "date",
          "description": "The due date of the task"
        },
        {
          "id": "fldCompleted",
          "name": "Completed",
          "type": "checkbox",
          "description": "Whether the task is completed"
        },
        {
          "id": "fldNotes",
          "name": "Notes",
          "type": "multilineText",
          "description": "Additional notes about the task"
        },
        {
          "id": "fldTags",
          "name": "Tags",
          "type": "multipleSelects",
          "description": "Tags associated with the task",
          "options": {
            "choices": [
              {
                "id": "selUrgent",
                "name": "Urgent",
                "color": "redLight2"
              },
              {
                "id": "selImportant",
                "name": "Important",
                "color": "orangeLight2"
              }
            ]
          }
        },
        {
          "id": "fldAttachments",
          "name": "Attachments",
          "type": "multipleAttachments",
          "description": "Files attached to the task"
        },
        {
          "id": "fldCreatedAt",
          "name": "Created At",
          "type": "createdTime",
          "description": "The time when the task was created"
        },
        {
          "id": "fldUpdatedAt",
          "name": "Updated At",
          "type": "lastModifiedTime",
          "description": "The time when the task was last updated"
        },
        {
          "id": "fldAssignedTo",
          "name": "Assigned To",
          "type": "multipleRecordLinks",
          "description": "Users assigned to the task",
          "options": {
            "linkedTableId": "tblUsers"
          }
        }
      ]
    },
    {
      "id": "tblUsers",
      "name": "users",
      "fields": [
        {
          "id": "fldEmail",
          "name": "Email",
          "type": "email"
        },
        {
          "id": "fldFullName",
          "name": "Full Name",
          "type": "singleLineText"
        },
        {
          "id": "fldPhone",
          "name": "Phone",
          "type": "phoneNumber",
          "description": "please include area code"
        }
      ]
    }
  ]
};
"
`;
